scalding-core/src/main/scala/com/twitter/scalding/graph/Graph.scala

```diff
@@ -0,0 +1,231 @@
+  def edges: TypedPipe[Edge[T, S]] = inputEdges
+  def vertices: TypedPipe[Vertex[T, Q]] = inputVertices
+
+  /**
+   * Returns a TypedPipe of edges with joined Vertex attributes.
+   */
+  def triplets: TypedPipe[EdgeTriplet[T, S, Q]] =
+    edges
+      .groupBy(_.source)
+      .join(vertices.groupBy(_.id))
+      .toTypedPipe
+      .map{ case (source, (edge, packet)) => (edge.dest, (edge, packet)) }
+      .join(vertices.groupBy(_.id))
+      .values
+      .map{ case ((edge, sourceVertex), destVertex) => EdgeTriplet(sourceVertex, destVertex, edge) }
+
+  /**
+   * Left join vertices with the graph vertices and generate a new Graph
+   */
+  def leftJoinVertices[U, VD2](other: TypedPipe[Vertex[T, U]])(mapFunc: (T, Q, Option[U]) => VD2): Graph[T, S, VD2] = {
+    val newVertices = vertices
+      .groupBy(_.id)
+      .leftJoin(other.groupBy(_.id))
+      .toTypedPipe
+      .map{ case (id, (vertex, data)) => Vertex(id, mapFunc(id, vertex.attr, data.map(_.attr))) }
+
+    new GraphUnfilteredEdges(edges, newVertices)
+  }
+
+  /**
+   * Inner join vertices with the graph vertices and generate a new Graph
+   */
+  def joinVertices[U, VD2](other: TypedPipe[Vertex[T, U]])(mapFunc: (T, Q, U) => VD2): Graph[T, S, VD2] = {
+    val newVertices = vertices
+      .groupBy(_.id)
+      .join(other.groupBy(_.id))
+      .toTypedPipe
+      .map{ case (id, (vertex, data)) => Vertex(id, mapFunc(id, vertex.attr, data.attr)) }
+
+    new Graph[T, S, VD2](edges, newVertices)
+  }
+
+  def mapVertices[A](map: Vertex[T, Q] => A): Graph[T, S, A] =
+    new Graph[T, S, A](edges, vertices.map{ vertex => Vertex(vertex.id, map(vertex)) })
+
+  def mapEdges[A](map: Edge[T, S] => A): Graph[T, A, Q] =
+    new Graph[T, A, Q](edges.map{ edge => edge.copy(attr = map(edge)) }, vertices)
+
+  def mapTriplets[A, B](map: EdgeTriplet[T, S, Q] => EdgeTriplet[T, A, B]): Graph[T, A, B] = {
+    val newTriplets = triplets.map(map)
+
+    new Graph[T, A, B](
+      newTriplets.map(_.edge),
+      newTriplets.flatMap(trip => List(trip.source, trip.dest)).distinct(Ordering.by(_.id)))
+  }
+
+  /**
+   * Collect only neighbor ids.
+   * Optionally sort by Vertex
+   */
+  def collectNeighborIds(sortNeighbors: Boolean = true): Graph[T, S, Neighbors[T]] = {
+    val nbrs = edges
+      .map { edge => (edge.source, edge.dest) }
+      .group
+      .mapGroup {
+        case (vert, nbrs) =>
+          val nbrsArray: Vertex[T, Neighbors[T]] = if (sortNeighbors) {
+            Vertex(vert, SortedNeighbors(nbrs.toArray.sorted)(ord))
+          } else {
+            Vertex(vert, UnsortedNeighbors(nbrs.toArray))
+          }
+
+          Iterator.single(nbrsArray)
+      }
+      .values
+
+    Graph(edges, nbrs)
+  }
+
+  /**
+   * Collect all Neighbors of an Vertex.
+   * Optionally sort by Vertex
+   */
+  def collectNeighbors(sortNeighbors: Boolean = true): Graph[T, S, Neighbors[Vertex[T, Q]]] = {
```

Comment:
can't the above call this one and just `mapVertices` and a `mapAttr` on `Vertex`?